src/Mvc/Mvc.Core/src/ApplicationParts/RelatedAssemblyAttribute.cs

```diff
@@ -91,6 +92,22 @@ public static IReadOnlyList<Assembly> GetRelatedAssemblies(Assembly assembly, bo
                         Resources.FormatRelatedAssemblyAttribute_AssemblyCannotReferenceSelf(nameof(RelatedAssemblyAttribute), assemblyName));
                 }
 
+                var relatedAssemblyName = new AssemblyName(attribute.AssemblyFileName);
+                Assembly relatedAssembly;
+                try
+                {
+                    // Perform a cursory check to determine if the Assembly has already been loaded
+                    // before going to disk. In the ordinary case, related parts that are part of
+                    // application's reference closure should already be loaded.
+                    relatedAssembly = Assembly.Load(relatedAssemblyName);
+                    relatedAssemblies.Add(relatedAssembly);
+                    continue;
+                }
+                catch (IOException)
+                {
+                    // The assembly isn't already loaded. Patience, we'll attempt to load it from disk next.
```

Comment:
First chance exceptions creeping into ASP.NET Core... Making me sad. Do we need this check?

Reply:
"simply try to load the assembly" - I meant basically remove the check and just always run the code below it - which will find the file on disk and load it (hopefully via something like `AssemblyLoadContext.LoadFromAssemblyPath`). Runtime has a cache which per-ALC stores all loaded assemblies - if it's asked to load the same file again, it will simply return the already loaded assembly. That check happens relatively early on and is pretty fast (no file IO if I remember correctly).